### PR TITLE
Recursive info field serialization

### PR DESCRIPTION
**Pull request by arcondello**
Created a more generic serialization for numpy arrays. This allows us to serialize the info field.

Slightly alters the serialization introduced in #511

### CODE REVIEW

File: dimod/sampleset.py [hunk 1]
```diff
@@ -1391,39 +1396,38 @@ def to_serializable(self, use_bytes=False, bytes_type=bytes):
...
             "num_variables": len(self.variables),
             "num_rows": len(self),
+            "sample_data": sample_data,
+            "sample_type": self.record.sample.dtype.name,
+
+            # vectors
+            "vectors": vectors,
+
+            # other
             "variable_labels": self.variables.to_serializable(),
```
Comment (randomir) [inline dimod/sampleset.py]: Do we need to handle NumPy scalars/primitives for variables here (e.g. `int32`)?

File: dimod/serialization/utils.py [hunk 2]
```diff
@@ -13,9 +13,88 @@
...
+    arr = arr.reshape(obj['shape'])  # makes a view generally, but that's fine
+    return arr
+
+
+def dfs_serialize_ndarray(obj, use_bytes=False, bytes_type=bytes):
+    """Looks through the object, serializing numpy arrays."""
+    if isinstance(obj, np.ndarray):
+        return serialize_ndarray(obj, use_bytes=use_bytes, bytes_type=bytes_type)
+    elif isinstance(obj, abc.Mapping):
+        return {key: dfs_serialize_ndarray(val) for key, val in obj.items()}
```
Comment (randomir) [inline dimod/serialization/utils.py]: Keys can be NumPy types too.

File: dimod/serialization/utils.py [hunk 3]
```diff
@@ -13,9 +13,88 @@
...
+    """
+    if obj['use_bytes']:
+        arr = np.frombuffer(obj['data'], dtype=obj['data_type'])
+    else:
+        arr = np.asarray(obj['data'], dtype=obj['data_type'])
+    arr = arr.reshape(obj['shape'])  # makes a view generally, but that's fine
+    return arr
+
+
+def dfs_serialize_ndarray(obj, use_bytes=False, bytes_type=bytes):
```
Comment (randomir) [inline dimod/serialization/utils.py]: Confusing name since also dicts and lists are serialized.

File: dimod/serialization/utils.py [hunk 4]
```diff
@@ -13,9 +13,88 @@
...
+    """Looks through the object, serializing numpy arrays."""
+    if isinstance(obj, np.ndarray):
+        return serialize_ndarray(obj, use_bytes=use_bytes, bytes_type=bytes_type)
+    elif isinstance(obj, abc.Mapping):
+        return {key: dfs_serialize_ndarray(val) for key, val in obj.items()}
+    elif isinstance(obj, abc.Sequence) and not isinstance(obj, str):
+        return list(map(dfs_serialize_ndarray, obj))
```
Comment (randomir) [inline dimod/serialization/utils.py]: List comprehensions are faster when you're not exploiting iter `map`.

File: dimod/serialization/utils.py [hunk 5]
```diff
@@ -13,9 +13,88 @@
...
+    return arr
+
+
+def dfs_serialize_ndarray(obj, use_bytes=False, bytes_type=bytes):
+    """Looks through the object, serializing numpy arrays."""
+    if isinstance(obj, np.ndarray):
+        return serialize_ndarray(obj, use_bytes=use_bytes, bytes_type=bytes_type)
+    elif isinstance(obj, abc.Mapping):
+        return {key: dfs_serialize_ndarray(val) for key, val in obj.items()}
+    elif isinstance(obj, abc.Sequence) and not isinstance(obj, str):
```
Comment (randomir) [inline dimod/serialization/utils.py]: Soon won't matter, but this won't handle unicode in Python 2. (you can use six.string_types, or something like that)

File: dimod/serialization/utils.py [hunk 6]
```diff
@@ -13,9 +13,88 @@
...
+        return list(map(dfs_serialize_ndarray, obj))
+    return obj
+
+
+def dfs_deserialize_ndarray(obj):
+    """Inverse of dfs_serialize_ndarray"""
+    if isinstance(obj, abc.Mapping):
+        if obj.get('type', '') == 'array':
```
Comment (randomir) [inline dimod/serialization/utils.py]: Out of curiosity (I've seen you used this pattern elsewhere) -- why use empty string for default (and not the default `None`)?

**Review by randomir — COMMENTED** [review]
Looks good.

I would just consider adding ints and floats to the serialization mix.

File: dimod/serialization/utils.py (hunk 2 above)
Comment (arcondello) [reply dimod/serialization/utils.py]: Not if it's JSON serializable. Unless numpy chars?

File: dimod/serialization/utils.py (hunk 6 above)
Comment (arcondello) [reply dimod/serialization/utils.py]: no real reason, matching type but None would work just as well.

File: dimod/sampleset.py (hunk 1 above)
Comment (arcondello) [reply dimod/sampleset.py]: Good question, definitely an edge case we should test/handle.

Comment (arcondello) [reply dimod/sampleset.py]: Handled by the Variables object https://github.com/dwavesystems/dimod/blob/368c8887948e62528a8369ea73744e2f584582d7/dimod/variables.py#L44

File: dimod/serialization/utils.py (hunk 2 above)
Comment (randomir) [reply dimod/serialization/utils.py]: Also numpy numbers.

File: dimod/serialization/utils.py (hunk 6 above)
Comment (randomir) [reply dimod/serialization/utils.py]: If not even better! :smiley_cat:

File: dimod/serialization/utils.py (hunk 2 above)
Comment (arcondello) [reply dimod/serialization/utils.py]: JSON needs string keys for objects. I don't think we should handle that case here.

Comment (arcondello) [reply dimod/serialization/utils.py]: Also seems to work for the numeric types out of the box
```
In [4]: json.dumps({np.float(0): 1})                                                                                                   
Out[4]: '{"0.0": 1}'
```

**Review by oneklc — APPROVED** [review]

File: dimod/serialization/utils.py (hunk 2 above)
Comment (randomir) [reply dimod/serialization/utils.py]: Yes, I know JSON requires string keys. But as you see, built-in json package serializes python numbers as strings. It fails with numpy though:
```
>>> json.dumps({np.int64(0): 1})
...
TypeError: keys must be a string
```

Comment (randomir) [reply dimod/serialization/utils.py]: I'm not saying I like casting arrays to strings, but this solution handles only ndarrays in values. I feel if we already are providing this service to users (serialization of numpy types), we should be little broader.

**Review by randomir — APPROVED** [review]

File: dimod/serialization/utils.py (hunk 2 above)
Comment (arcondello) [reply dimod/serialization/utils.py]: It's a simple enough change so will do. However, I don't think it is in scope for dimod to handle serialization of numpy objects in general. So far we are only providing it as an addition to the serialization of dimod objects. If we want to go beyond that this code will need to be a lot more sophisticated.

Comment (randomir) [reply dimod/serialization/utils.py]: I agree. Actually, if you remember, I argued even more than that -- BQM/SampleSet serialization (for SAPI encoding) does not belong in dimod at all.

Comment (arcondello) [reply dimod/serialization/utils.py]: I think serialization of dimod objects makes sense in dimod. The numpy objects that happen to be stored less so.

Comment (randomir) [reply dimod/serialization/utils.py]: (But as long as we have it already, and we're now adding NumPy support too, I'm saying let's just cover simple edge cases to minimize gotcha moments.)
